### ISSUE TITLE
is it still in progress ?

### ISSUE THREAD

**Issue by Rover-Yu** [linux-okernel/linux-okernel#4]
Hi, 

This project is very interesting, I ever have similar ideas with you.  Are you still in progress ?

And, I have question, why you don't use the MMU based isolation like nested kernel ?

Thanks
Yu

**Comment by t-koulouris** [comment]
Hi,

Thanks for your interest.
Yes, this is still under active development. Have a look at the project description here: https://github.com/linux-okernel/linux-okernel/blob/linux-okernel/Documentation/okernel.txt

Does this answer your question? Could you please clarify your idea?

Thanks
Theo

**Comment by adrianlshaw** [comment]
@Rover-Yu IIRC Nested Kernel (by Dautenhahn) needs to control which privileged instructions can be used by outer kernel code. It has to do this by scanning the pages for problematic instructions (e.g. MMU operations) and subsequently modify them. This is tricky to do correctly in software. This repo uses ISA virtualisation extensions, which can trap the sensitive instructions for you and thus can lead to better performance and correctness.

**Comment by Rover-Yu** [comment]
You started a so interesting project for container isolation, instead of just followed unikernel idea :) 

It seem that your hacked dune kernel module is not open sourced at github ?

Basically, my idea is to use pure MMU based isolation like nested kernel, but I want to enable multiple "managed/lightweight (Linux) kernel" to run non root containers at same time.

One of my ideas was to use EPT (or stage 2 translation table in ARMv8a) to isolate "container kernel" too.  I give up it since I think that this solution should have similar drawbacks with regular VM, .e.g. poor I/O performance. What do your think about these possible challenges ?

**Comment by Rover-Yu** [comment]
@adrianlshaw you are right ! to keep these invariants in nested kernel is not trivial , but I think that we can make a try under Linux :)

**Comment by adrianlshaw** [comment]
The code in this repo already uses EPT and will be a basis for protection. ARM support could be added with stage 2 translation, like you said. I/O performance is an issue if you have to do any resource multiplexing. You could assign individual devices to a particular outer kernel for full performance using IOMMU, but I don't think there's any code here that takes advantage of that (yet).

**Comment by adrianlshaw** [comment]
@Rover-Yu The code is not a kernel module, it has been integrated into the scheduler and other places. Check the different branches in the repo, like 'linux-okernel', not the master branch. The code is in  **arch/x86/okernel**. 

The userspace tools for creating an outer kernel are in a separate repo.
